deploy/s3.sh

```diff
@@ -0,0 +1,189 @@
+# Also make sure your default region is correct, otherwise, override with $S3_REGION
+
+########  Public functions #####################
+
+#domain keyfile certfile cafile fullchain
+s3_deploy() {
+  _cdomain="$1"
+  _ckey="$2"
+  _ccert="$3"
+  _cca="$4"
+  _cfullchain="$5"
+  
+  if [ -z "$S3_BUCKET" ] ; then
+    _err "You haven't specified the bucket name yet."
+    _err "Please set it via export and try again."
+    _err "e.g. export S3_BUCKET=acme"
+    return 1
+  fi
+
+  if [ -z "$AWS_PROFILE" ]; then
+    AWS_PROFILE="default"
+  fi
+
+  if ! _exists aws; then
+    _debug "AWS CLI not installed, defaulting to curl method"
+    _aws_cli_installed=0
+  else
+    _debug "AWS CLI installed, defaulting ignoring curl method"
+    _aws_cli_installed=1
+    S3_REGION="$(aws configure get region --profile ${AWS_PROFILE})"
+  fi
+
+  if [ "$_aws_cli_installed" -eq "0" ] && ([ -z "$AWS_ACCESS_KEY_ID" ] || [ -z "$AWS_SECRET_ACCESS_KEY" ]); then
+    _err "AWS_ACCESS_KEY_ID or AWS_SECRET_ACCESS_KEY not set."
+    _err "Please set them via export, or use the aws-cli."
+    return 1
+  fi
+
+  if [ -z "$S3_REGION" ]; then
+    S3_REGION="us-east-1"
+  fi
+
+  # Save s3 options if it's succesful (First run case)
+  _saveaccountconf S3_BUCKET "$S3_BUCKET"
+  _saveaccountconf S3_REGION "$S3_REGION"
+  _saveaccountconf AWS_PROFILE "$AWS_PROFILE"
+
+  _debug _cdomain "$_cdomain"
+  _debug _ckey "$_ckey"
+  _debug _ccert "$_ccert"
+  _debug _cca "$_cca"
+  _debug _cfullchain "$_cfullchain"
+  _debug S3_BUCKET "$S3_BUCKET"
+  _debug AWS_PROFILE "$AWS_PROFILE"
+  _secure_debug AWS_ACCESS_KEY_ID "$AWS_ACCESS_KEY_ID"
+  _secure_debug AWS_SECRET_ACCESS_KEY "$AWS_SECRET_ACCESS_KEY"
+  
+  # REMOVE BEFORE COMMIT, ONLY FOR DEBUGGING
+  _aws_cli_installed=1
+
+  _info "Deploying certificate to s3 bucket: $S3_BUCKET in $S3_REGION"
+  
+  if [ "$_aws_cli_installed" -eq "0" ]; then
+    _debug "deploying with curl method"
+  else
+    _debug "deploying with aws cli method"
+  fi
+
+  # private
+  _deploy_to_bucket "$_ckey" "$_cdomain/$_cdomain.key"
+  # public
+  _deploy_to_bucket "$_ccert" "$_cdomain/$_cdomain.cer"
+  # ca
+  _deploy_to_bucket "$_cca" "$_cdomain/ca.cer"
+  # fullchain
+  _deploy_to_bucket "$_cfullchain" "$_cdomain/fullchain.cer"
+
+  return 0
+
+}
+
+####################  Private functions below ##################################
+
+_deploy_to_bucket() {
+  if [ "$_aws_cli_installed" -eq "0" ]; then
+    _deploy_with_curl "$1" "$2"
+  else
+    _deploy_with_awscli "$1" "$2"
+  fi
+}
+
+_deploy_with_awscli() {
+  file="$1"
+  prefix="$2"
+  aws s3 cp "$file" s3://"$S3_BUCKET"/"$prefix" --region "$S3_REGION" --profile "$AWS_PROFILE"
+}
+
+_deploy_with_curl() {
+
+  file="${1}"
+  bucket="${S3_BUCKET}"
+  prefix="${2}"
+  region="${S3_REGION}"
+  acl="private"
+  timestamp="$(date -u "+%Y-%m-%d %H:%M:%S")"
+  signed_headers="date;host;x-amz-acl;x-amz-content-sha256;x-amz-date"
+
+  if [ "$(uname)" = "Darwin" ]; then
```

Comment:
please take a look at the `aws_rest()` function:
https://github.com/Neilpang/acme.sh/blob/master/dnsapi/dns_aws.sh#L156

It may help construct the aws headers.